### PR TITLE
commits for PR review

### PR DESCRIPTION
**Pull request by numb3r3**
This brank is only the copy of `main` branch. 

**THIS PR IS ONLY FOR CODE REVIEW. DON'T MERGE IT.**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,28 @@
 # FaissIndexer
 
-A similarity search based on Faiss and LMDB
+A similarity search indexer based on Faiss.
```
Comment (cristianmtr) [inline README.md:3]: Expand this to mention flexible storage

File: executor/indexer.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        if self._vec_indexer:
+            self._vec_indexer.reset()
+        self._buffer_indexer.clear()
+        self._metas = {'doc_ids': [], 'doc_id_to_offset': {}, 'delete_marks': []}
+        self._bloom = self._init_bloom()
+
+    @requests(on='/status')
+    def status(self, **kwargs):
+        """Return the status of the indexer."""
+        status = Document(tags={'db_stat': self._kv_db.stat})
```
Comment (cristianmtr) [inline executor/indexer.py:223]: Does this actually work? Doesn't it need to be appended as a match? Maybe add an integration test with a Flow

File: executor/indexer.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        self._metas = {'doc_ids': [], 'doc_id_to_offset': {}, 'delete_marks': []}
+
+        self._bloom = self._init_bloom()
+
+        # the kv_db is the storage backend for documents
+        self.logger.info(f'Using "{storage_backend}" as the storage backend')
+        self._kv_db = StorageFactory.open(storage_backend, db_path=storage_path)
+
+        # the buffer_indexer is created for incremental updates
+        self._buffer_indexer = DocumentArray()
```
Comment (cristianmtr) [inline executor/indexer.py:69]: Scaling this across shards/replicas will be a challenge

**Review by cristianmtr — CHANGES_REQUESTED** [review]
initial things. Will review more once we talk about it in the Hub meeting tomorrow

File: executor/indexer.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        super().__init__(*args, **kwargs)
+        self.logger = JinaLogger(self.__class__.__name__)
+
+        self.index_key = index_key
+        self.metric = metric
+
+        workspace = Path(self.workspace)
+        storage_path = str(workspace / 'lmdb_docs')
+
+        self._metas = {'doc_ids': [], 'doc_id_to_offset': {}, 'delete_marks': []}
```
Comment (tadejsv) [inline executor/indexer.py:60]: You are using two dictionaries, `doc_ids` and `doc_id_to_offset`. The problem that you have here is that you have to manually make sure they are in sync. Consider using `bidict` package for that.

File: executor/indexer.py (hunk 3 above)
Comment (numb3r3) [reply executor/indexer.py:69]: yes, good catch. Now, the sharding problem has not been considered now. So far we don't have a perfect solution for it. Maybe we can first collect some feedbacks about `FaissPSQL`. If the sharding/replica consistency solution in `FaisPSQL` can work well. Then, we can directly copy the ideas from `FaissPSQL`.

File: executor/indexer.py (hunk 4 above)
Comment (numb3r3) [reply executor/indexer.py:60]: thanks for your suggestion. `bidict` is perfect, I would like to use it.

Comment (numb3r3) [reply executor/indexer.py:60]: BTW, May I ask for your help to make this update (if you are available)? https://github.com/jina-ai/executor-faissindexer/issues/8

File: executor/indexer.py (hunk 2 above)
Comment (numb3r3) [reply executor/indexer.py:223]: you are right. We need an integration test to guarantee all endpoints work as expected. I created an issue https://github.com/jina-ai/executor-faissindexer/issues/9 to track this task.

File: executor/indexer.py (hunk 4 above)
Comment (tadejsv) [reply executor/indexer.py:60]: Sure, I can work on it 👍

**Pull request closed without merging**